029-across-step/proposal.md

```diff
@@ -0,0 +1,277 @@
+# `across` step
+
+This proposal introduces a mechanism by which a build plan can be executed
+multiple times with different `((vars))`, in service of build matxies, pipeline
+matrixes, and - building on [`((var))` sources][var-sources-rfc] - dynamic
+variations of both.
+
+
+## Motivation
+
+* Support dynamic multi-branch workflows:
+  [concourse/concourse#1172][multi-branch-issue]
+* Support build matrixes and pipeline matrixes.
+* Replace a common, and very flawed, usage pattern of `version: every`, which
+  has proven to be very complicated to support.
+
+
+## Proposal
+
+The `across` step modifier is given a list containing vars and associated
+values to set when executing the step. The step will be executed across all
+combinations of var values.
+
+### Static var values
+
+The set of values can be static, as below:
+
+```yaml
+task: unit
+vars: {go_version: ((.:go_version))}
+across:
+- var: go_version
+  values: [1.12, 1.13]
+```
+
+This will run the `unit` task twice: once with `go_version` set to `1.12`, and
+again with `1.13`.
+
+### Dynamic var values from a var source
+
+Rather than static values, var values may be pulled from a var source
+dynamically:
+
+```yaml
+var_sources:
+- name: booklit-prs
+  type: github-prs
+  config:
+    repository: vito/booklit
+    access_token: # ...
+
+plan:
+- set_pipeline: pr
+  instance_vars: {pr_number: ((.:pr.number))}
+  across:
+  - var: pr
+    source: booklit-prs
+```
+
+The above example will run the `set_pipeline` step across the set of all GitHub
+PRs, returned through a `list` operation on the `booklit-prs` var source.
```

Comment:
This may not be relevant discussion to this RFC, but I'm a bit confused about how `var_sources` work in the context of PRs. It seems to me like there's a separation between credential manager `var_sources` and PR-like `var_sources`.

It don't really see the use case to run a plan `across` a list of credentials from Vault, nor do I see the use case to access a single PR via a `var_source` (e.g. `((booklit-prs:1234))` to get some info about PR#1234? - based on your comment here https://github.com/concourse/rfcs/pull/37#discussion_r428195212, it sounds like you'd want to `get` a PR via the resource interface instead)

Reply:
The unifying theory between var sources is only this: they discover a set of things (credential paths, branches, PRs) and have a way to fetch additional data for each thing (credential value, branch/PR metadata). It's not really suggesting that they all have the same usage - only that they fit the same shape, and Concourse has primitives that work with those shapes.

Who knows: maybe someone does have a use case for using `across` with a credential manager? Perhaps they want to monitor all credentials to detect whether any of them have been accidentally checked in to the codebase, or whether they are due to be rotated.

It's true that you would use a `get` step to fetch the PR's bits, but you may also want to get the attributes of a PR in order to, say, periodically check that the PR assignees have been active. This could be done through a job in the PR pipeline that fetches the PR var:

```yaml
jobs:
- name: ping-assignees
  plan:
  - get_var: time
    source: daily
    trigger: true
  - get_var: pr
    path: ((pr_number))
    source: booklit-prs
  - run: check-activity
    type: pr-monitor
    params:
      assignees: ((.:pr.assignees))
      last_activity_at: ((.:pr.last_activity_at))
      last_active_user: ((.:pr.last_active_user))
```

note: I cheated there - `get_var` doesn't have a `path:` - but it probably should, otherwise you'd have wacky syntax like `((.:((pr_number)).assignees))`.

Reply:
Makes sense, thanks for the clarification! Those are some pretty cool use cases, and as you say, I'm sure there's plenty more people will come up with.

Excited for more of these v10 ideas to land!

Reply:
Sorry, one last thought around this - given that secrets from credential managers aren't all necessarily the same shape, do you think it makes sense to extend the var interpolation syntax to provide a way to coerce any var into a string? For instance:

```yaml
task: detect-leaked-secret
vars:
  secret: ((.:secret | json)) # (or `| yaml`)
across:
- var: secret
  source: vault
```

Since otherwise the job would error if any secret isn't a string

Reply:
@aoldershaw Hmm, it's true that with `vars:` you would likely need some sort of marshalling because even though `vars:` supports complex structures, the task config template wouldn't really know what to do with it. I think you could avoid marshalling/unmarshalling values if you use a Prototype and pass the value along through `params:` instead, where it would become part of the JSON request structure. I would almost prefer this approach so that folks write more prototypes instead of tasks.

Also, one tricky detail with credential scanning: `across` would get the vars from `list`, which would not include the sensitive values themselves. I think credential managers should omit sensitive values from the `list` call so that we can safely list available credentials in the web UI (to authorized parties, anyway). Doing so would really help with discoverability: "what was that credential called again?" -> navigate to the var source -> use `(())` syntax. So you would need a `get_var` step somewhere to fetch the actual credential.